daml-integration-kit: release PRE-ALPHA version in SDK

meiersi-da opened the pull request:
See the documentation for the DAML Integration Kit for the definition of PRE-ALPHA.

This PR is ready for review.

Contributes to: #135 

### Pull Request Checklist

- [x] Read and understand the [contribution guidelines](https://github.com/digital-asset/daml/blob/master/CONTRIBUTING.md)
- [x] Include appropriate tests
- [x] Set a descriptive title and thorough description
- [x] Add a reference to the [issue this PR will solve](https://github.com/digital-asset/daml/issues), if appropriate
- [x] Add a line to the [release notes](https://github.com/digital-asset/daml/blob/master/docs/source/support/release-notes.rst), if appropriate

NOTE: CI is not automatically run on non-members pull-requests for security
reasons. The reviewer will have to comment with `/AzurePipelines run` to
trigger the build.

mziolekda commented on docs/source/daml-integration-kit/index.rst:
```diff
@@ -0,0 +1,329 @@
+  contain holes, which are usually linked to GitHub issues.
+
+ALPHA (`tracking GitHub issue <https://github.com/digital-asset/daml/issues/658>`__)
+  Once we reach ALPHA status, the DAML Integration Kit is ready to be used by
```
I would suggest slightly less self-debilitating tone:
"Once ALPHA status is reached, the DAML Integration Kit is ready to be used by
third-parties willing to accept following caveats:"

mziolekda commented on docs/source/daml-integration-kit/index.rst:
```diff
@@ -0,0 +1,329 @@
+  courageous third-parties. If you are one of these pioneers, then we are glad
+  to have you on-board, provided your are OK with the following caveats:
+
+  - The architecture includes everything required to run DAML Applications in
```
This section is not so clear to me. Isn't the problem that several APIs are missing at the moment, thus the solutions to problems such as package upload, party on-boarding etc. are necessarily proprietary. The fact that they cannot be tested in a uniform manner is a consequence of that

meiersi-da replied:
Thx. reformulated to talk about introducing APIs

mziolekda commented on docs/source/daml-integration-kit/index.rst:
```diff
@@ -0,0 +1,329 @@
+``api-server-damlonx.jar`` (`source code <https://github.com/digital-asset/daml/blob/master/ledger/api-server-damlonx/src/main/scala/com/daml/ledger/api/server/damlonx/Server.scala>`__)
+  contains code that implements a DAML Ledger API
+  server given implementations of the interfaces in ``participant-state.jar``.
+``daml-engine.jar`` (`source code <https://github.com/digital-asset/daml/blob/master/daml-lf/engine/src/main/scala/com/digitalasset/daml/lf/engine/Engine.scala>`__)
```
Some readers might be looking for clues how daml engine fits into the overall architecture depicted on the diagram below. Maybe we should show how it fits into the LedgerApiServer.

meiersi-da replied:
Thx. Added a comment.

meiersi-da commented:
@bethaitman : I'd highly value a review from you. Feel free to edit the sources directly and push your changes as a commit. I'll read through that commit to learn from your changes and strive to do more of them right from the get go next time ;-)

@gleber-da @dajmaki @timw-da please review to swap in the content in case you have not yet done so.

bethaitman commented on docs/source/daml-integration-kit/index.rst:
```diff
@@ -0,0 +1,371 @@
+
+1. We provide Scala libraries for validating DAML transactions and serving the
+   Ledger API given implementations of two specific interfaces. See
+   :ref:`below <integration-kit_library_overview>` for an overview of these
```
That ref doesn't seem to work for me. 

In general, better to say "the library overview below" rather than just "below" - making link text meaningful is generally helpful

meiersi-da replied:
done

bethaitman commented on docs/source/daml-integration-kit/index.rst:
```diff
@@ -0,0 +1,371 @@
+`<X>-daml-validator` as explained above. We provide two kinds of code to
+simplify their implementation.
+
+1. We provide Scala libraries for validating DAML transactions and serving the
```
This isn't an ordered list, it can just be bullet points

meiersi-da replied:
adapted.

bethaitman commented on docs/source/daml-integration-kit/index.rst:
```diff
@@ -0,0 +1,371 @@
+`<X>-daml-validator` as explained above. We provide two kinds of code to
+simplify their implementation.
+
+1. We provide Scala libraries for validating DAML transactions and serving the
```
```suggestion
- Scala libraries for validating DAML transactions and serving the
```

meiersi-da replied:
Done.

bethaitman commented on docs/source/daml-integration-kit/index.rst:
```diff
@@ -0,0 +1,371 @@
+   :ref:`below <integration-kit_library_overview>` for an overview of these
+   libraries.
+
+2. We provide the `digital-asset/daml-on-x-example <https://github.com/digital-asset/daml-on-x-example>`__ GitHub repository,
```
```suggestion
- The `digital-asset/daml-on-x-example <https://github.com/digital-asset/daml-on-x-example>`__ GitHub repository,
```

meiersi-da replied:
Done

bethaitman commented on docs/source/daml-integration-kit/index.rst:
```diff
@@ -0,0 +1,371 @@
+   libraries.
+
+2. We provide the `digital-asset/daml-on-x-example <https://github.com/digital-asset/daml-on-x-example>`__ GitHub repository,
+   which contains a complete example of a DAML Ledger backed by an in-memory
```
Even better, switch these two bits of the sentence around: A complete example of a DAML ledger backed by key-value stored, in the 'daml-on-x' GitHub repository.

meiersi-da replied:
nice! :muscle:

bethaitman commented on docs/source/daml-integration-kit/index.rst:
```diff
@@ -0,0 +1,371 @@
+
+   For ledgers where data is shared between all participant nodes, we
+   recommend using this example as a starting point for implementing your
+   server and validator. For ledgers with stronger privacy models, this example
```
I would add a second paragraph break before "For ledgers with stronger privacy models"

meiersi-da replied:
:+1:

bethaitman commented on docs/source/daml-integration-kit/index.rst:
```diff
@@ -0,0 +1,371 @@
+   implement :doc:`DAML's privacy model </concepts/ledger-model/ledger-privacy>`
+   at the ledger level instead of just at the Ledger API level.
+
+Implementing your own server and validator requires a significant amount
```
Should this paragraph be under the next heading?

meiersi-da replied:
moved.

bethaitman commented on docs/source/daml-integration-kit/index.rst:
```diff
@@ -0,0 +1,371 @@
+implement it.
+
+
+Getting started
```
```suggestion
Prerequisite knowledge
```

bethaitman commented on docs/source/daml-integration-kit/index.rst:
```diff
@@ -0,0 +1,371 @@
+   :ref:`below <integration-kit_library_overview>` for an overview of these
+   libraries.
+
+2. We provide the `digital-asset/daml-on-x-example <https://github.com/digital-asset/daml-on-x-example>`__ GitHub repository,
```
You should definitely mention that this doesn't exist yet. Maybe in the introductory paragraph above say "We _will_ provide"

meiersi-da replied:
clarified.

bethaitman commented on docs/source/daml-integration-kit/index.rst:
```diff
@@ -0,0 +1,371 @@
+   **TODO (ALPHA):** create this example repository
+   (`issue <https://github.com/digital-asset/daml/issues/139>`__)
+
+2. Read the example code jointly with
```
Given this, I think it makes sense to have these instructions _first_, then the architectural overview _later_. It could even be its own page.

bethaitman commented on docs/source/daml-integration-kit/index.rst:
```diff
@@ -0,0 +1,371 @@
+Implementing your server and validator
+======================================
+
+Each `X` ledger requires at least the implementation of a specific
```
I think this content could do with its own heading - something like `Resources to help with ledger development`- and it can go further down, after the 'getting started' and 'deciding on the architecture' instructions. The reason being, those two sections are effectively acting as an index.

bethaitman commented on docs/source/daml-integration-kit/index.rst:
```diff
@@ -0,0 +1,371 @@
+   `digital-asset/daml <https://github.com/digital-asset/daml>`__ repository
+   to add your ledger to the `DAML Ledger overview`_.
+
+Use the feedback form on this documentation page or GitHub issues on the
```
This seems like a slightly odd place to ask for feedback - it's a bit randomly in the middle of the page.

meiersi-da replied:
adjusted. wanted them to know how to reach out for help.

meiersi-da commented on docs/source/daml-integration-kit/index.rst:
```diff
@@ -0,0 +1,371 @@
+.. Copyright (c) 2019 Digital Asset (Switzerland) GmbH and/or its affiliates. All rights reserved.
+.. SPDX-License-Identifier: Apache-2.0
+
+.. _integration-kit:
```
removed

meiersi-da commented on docs/source/daml-integration-kit/index.rst:
```diff
@@ -0,0 +1,371 @@
+implement such a server from scratch following the above documentation.
+Practically, we though strongly recommend to use our integration kit and follow
+the guides below for
+:ref:`implementing <integration-kit_implementing>`,
```
done

meiersi-da commented on docs/source/daml-integration-kit/index.rst:
```diff
@@ -0,0 +1,371 @@
+:ref:`deploying <integration-kit_deploying>`,
+:ref:`testing <integration-kit_testing>`, and
+:ref:`benchmarking <integration-kit_benchmarking>` your own DAML Ledger
+server. This way you can focus on your distributed-ledger or database of
```
Thx. Formulated a bit differently, but used the reference to the guides.

meiersi-da commented on docs/source/daml-integration-kit/index.rst:
```diff
@@ -0,0 +1,371 @@
+implementing the DAML Ledger API.
+
+
+.. _integration-kit_status-and_roadmap:
```
OK. removed.

meiersi-da commented on docs/source/daml-integration-kit/index.rst:
```diff
@@ -0,0 +1,371 @@
+
+  Before reaching BETA status, we expect to have
+
+  - hardened our test tooling,
```
Increased coverage, but I don't want to imply that it is weak at the moment. Leaving it intentionally vague.

meiersi-da commented on docs/source/daml-integration-kit/index.rst:
```diff
@@ -0,0 +1,371 @@
+
+  Before reaching BETA status, we expect to have
+
+  - hardened our test tooling,
```
applying these changes manually

meiersi-da commented on docs/source/daml-integration-kit/index.rst:
```diff
@@ -0,0 +1,371 @@
+explains what we mean by these different statuses, and what's missing to
+progress.
+
+PRE-ALPHA (current status)
```
Jup. Always running with the preview script.

meiersi-da commented on docs/source/daml-integration-kit/index.rst:
```diff
@@ -0,0 +1,371 @@
+
+.. original: https://www.lucidchart.com/invitations/accept/69799877-4e80-444d-96a3-3e90814e94df
+
+We assume that the `X` ledger allows entities to participate in the
```
Yes. deliberate use. The styling was too jarring, and it worked OK with the flow.

meiersi-da commented on docs/source/daml-integration-kit/index.rst:
```diff
@@ -0,0 +1,371 @@
+ledger `X`. This can be a proper distributed ledger or also just a database.
+The goal of this implementation is to allow multiple DAML applications, which are potentially run by
+different entities, to execute multi-party workflows using the ledger `X`. We
+illustrate this in the following diagram.
```
OK. Let me see what I can do. Just saw that our doc styleguide does not allow "we" the writers to speak, which is the style I've used in my papers.

meiersi-da commented:
@mziolekda thanks for your feedback. I've addressed all your comments. Let me know if you have more.

@bethaitman : I've addressed the comments that were simple to address. I'm now looking into how to improve the structure to satisfy the other comments you've had.

meiersi-da commented on docs/source/daml-integration-kit/index.rst:
```diff
@@ -0,0 +1,371 @@
+  That said: we highly value your feedback and input to direct our attention
+  to where it provides the most value!
+
+  Below is the list of known missing work to reach ALPHA status. You can
```
Agreed. Fixed with reference to https://github.com/digital-asset/daml/milestone/4

meiersi-da commented on docs/source/daml-integration-kit/index.rst:
```diff
@@ -0,0 +1,371 @@
+  - Some of our code might be fresh off the press and might therefore have a
+    higher rate of bugs.
+
+  That said: we highly value your feedback and input to direct our attention
```
Fixed.

gleber-da commented on docs/source/daml-integration-kit/index.rst:
```diff
@@ -0,0 +1,375 @@
+  is the collection of gRPC
+  services that you would like your `daml-on-<X>-server` to provide.
+``<X> services``
+  are the services provided by our underlying ledger,
```
`our underlying ledger` is confusing. My understanding is that all `X` things are provided by implementors. 
```suggestion
  are the services which underlie your ledger,
```

meiersi-da replied:
done

gleber-da commented on docs/source/daml-integration-kit/index.rst:
```diff
@@ -0,0 +1,375 @@
+  backed by the ``<X> services``. You need to implement this for your DAML on
+  `X` ledger.
+``WriteService`` (`source code <https://github.com/digital-asset/daml/blob/master/ledger/participant-state/src/main/scala/com/daml/ledger/participant/state/v1/WriteService.scala>`_)
+  is an interface abstracting over the ability to submit
```
```suggestion
  is an interface abstracting over the mechanism to submit
```
The interface abstracts away (or over) the mechanism which implements X, while provides an ability to do X at higher level.

meiersi-da replied:
done

gleber-da commented on docs/source/daml-integration-kit/index.rst:
```diff
@@ -0,0 +1,375 @@
+  is an interface abstracting over the ability to submit
+  DAML transactions to the underlying `X` ledger via a participant node.
+``ReadService`` (`source code <https://github.com/digital-asset/daml/blob/master/ledger/participant-state/src/main/scala/com/daml/ledger/participant/state/v1/ReadService.scala>`__)
+  is an interface abstracting over the ability to subscribe to
```
```suggestion
  is an interface abstracting over the mechanism to subscribe to
```

meiersi-da replied:
done

gleber-da commented:
@meiersi-da Prefer to avoid merging `origin/master` into your PR branches, as per current DAML team recommendations. It breaks merge queue tooling and creates messy git history, especially in a monorepo environment with large number of collaborators and PRs. Instead do `git pull --rebase origin master && git push -f origin` to rebase your branch on top of the master.

gleber-da commented:
@meiersi-da I will rebase and merge my docs PR on top of master after this PR lands. I dislike having multiple people contribute to a single PR, since it tends to make PRs messier and bigger than they ideally should be.

meiersi-da commented:
Ack wrt merge. Will do the rebase approach.

@bethaitman currently holds the edit token. She offered to make a pass on the structure. I leave it to her to rebase or squash, and consider accepting some of your changes.

Happy to do them also when I get the edit token back.

bethaitman commented:
I have made my edits - handing back the edit token!

meiersi-da commented:
Thanks a lot @bethaitman-da -- looks much better. I did some minor polishing and believe we're good to merge. Could you greenlight?

meiersi-da commented on docs/scripts/live-preview.sh:
```diff
@@ -35,6 +35,8 @@ tar -zxf ../../bazel-genfiles/docs/da_theme.tar.gz -C $BUILD_DIR/theme
 cp ../../LICENSE ../source
 cp ../../NOTICES ../source
 
+#
```
ignored. not my file.
